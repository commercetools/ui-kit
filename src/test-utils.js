import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider, addLocaleData } from 'react-intl';
import { Router } from 'react-router-dom';
import { createMemoryHistory } from 'history';
import en from 'react-intl/locale-data/en';
import de from 'react-intl/locale-data/de';
import es from 'react-intl/locale-data/es';
import messages from '../i18n/core.json';

addLocaleData(en);
addLocaleData(de);
addLocaleData(es);

const customRender = (
  node,
  {
    locale = 'en', // react-router
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
    ...rtlOptions
  } = {}
) => ({
  ...render(
    <IntlProvider locale={locale} messages={messages}>
      <Router history={history}>{node}</Router>
    </IntlProvider>,
    rtlOptions
  ),
  // adding `history` to the returned utilities to allow us
  // to reference it in our tests (just try to avoid using
  // this to test implementation details).
  history,
});

// re-export everything
export * from 'react-testing-library';

// override render method
export { customRender as render };
