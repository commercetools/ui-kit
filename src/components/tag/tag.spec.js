import React from 'react';
import { render } from '../../test-utils';
import Tag from './tag';

it('should render children', () => {
  const { container } = render(<Tag>Bread</Tag>);
  expect(container).toHaveTextContent('Bread');
});

it('should call onClick when clicked', () => {
  const onClick = jest.fn();
  const { getByText } = render(<Tag onClick={onClick}>Bread</Tag>);
  getByText('Bread').click();
  expect(onClick).toHaveBeenCalledTimes(1);
});

it('should call onRemove when remove icon is clicked', () => {
  const onRemove = jest.fn();
  const { getByLabelText } = render(<Tag onRemove={onRemove}>Bread</Tag>);
  getByLabelText('Remove').click();
  expect(onRemove).toHaveBeenCalledTimes(1);
});

it('should call the right handler when onClick and onRemove are provided', () => {
  const onClick = jest.fn();
  const onRemove = jest.fn();
  const { getByText, getByLabelText } = render(
    <Tag onRemove={onRemove} onClick={onClick}>
      Bread
    </Tag>
  );

  // clicking the Tag
  getByText('Bread').click();
  expect(onClick).toHaveBeenCalledTimes(1);
  expect(onRemove).not.toHaveBeenCalled();

  // clicking the remove button
  getByLabelText('Remove').click();
  expect(onRemove).toHaveBeenCalledTimes(1);
  expect(onClick).toHaveBeenCalledTimes(1);
});

describe('when disabled', () => {
  it('should not call onClick when clicked', () => {
    const onClick = jest.fn();
    const { getByText } = render(
      <Tag onClick={onClick} isDisabled={true}>
        Bread
      </Tag>
    );
    getByText('Bread').click();
    expect(onClick).not.toHaveBeenCalled();
  });

  it('should not call onRemove when clicked', () => {
    const onRemove = jest.fn();
    const { getByLabelText } = render(
      <Tag onRemove={onRemove} isDisabled={true}>
        Bread
      </Tag>
    );
    getByLabelText('Remove').click();
    expect(onRemove).not.toHaveBeenCalled();
  });
});

describe('when linkTo is set', () => {
  it('should redirect when clicked', () => {
    const { getByText, history } = render(<Tag linkTo="/foo">Bread</Tag>);
    getByText('Bread').click();

    expect(history.location.pathname).toBe('/foo');
  });

  it('should still call onClick when clicked', () => {
    const onClick = jest.fn();
    const { getByText, history } = render(
      <Tag onClick={onClick} linkTo="/foo">
        Bread
      </Tag>
    );
    getByText('Bread').click();
    expect(onClick).toHaveBeenCalled();

    expect(history.location.pathname).toBe('/foo');
  });

  it('should not redirect when removed', () => {
    const onRemove = jest.fn();
    const { getByLabelText, history } = render(
      <Tag onRemove={onRemove} linkTo="/foo">
        Bread
      </Tag>
    );

    getByLabelText('Remove').click();

    // ensure "onRemove" is stil called
    expect(onRemove).toHaveBeenCalled();

    // ensure the pathname is not "/foo", otherwise a redirect would have
    // happend
    expect(history.location.pathname).toBe('/');
  });

  it('should not redirect when disabled', () => {
    const { getByText, history } = render(
      <Tag linkTo="/foo" isDisabled={true}>
        Bread
      </Tag>
    );

    getByText('Bread').click();

    // ensure the pathname is not "/foo", otherwise a redirect would have
    // happend
    expect(history.location.pathname).toBe('/');
  });
});
