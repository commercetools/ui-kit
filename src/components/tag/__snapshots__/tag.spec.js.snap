// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tag /> rendering when \`linkTo\` is not set should output correct tree 1`] = `
<Horizontal
  constraint="scale"
>
  <div
    className="wrapper wrapperTypeNormal"
  >
    <TagNormalBody
      isDisabled={false}
      onClick={[MockFunction]}
      onRemove={[MockFunction]}
      type="normal"
    >
      Icecream we scream!
    </TagNormalBody>
    <AccessibleButton
      buttonAttributes={Object {}}
      className="removeWrapper removeWrapperTypeNormal"
      isDisabled={false}
      isToggleButton={false}
      label="Remove"
      onClick={[MockFunction]}
      type="button"
    >
      <CloseBoldIcon
        size="medium"
      />
    </AccessibleButton>
  </div>
</Horizontal>
`;

exports[`<Tag /> rendering when \`linkTo\` is set should output correct tree 1`] = `
<Horizontal
  constraint="scale"
>
  <div
    className="wrapper wrapperTypeNormal"
  >
    <TagLinkBody
      isDisabled={false}
      linkTo="/foo"
      onClick={[MockFunction]}
      type="normal"
    >
      Icecream we scream!
    </TagLinkBody>
  </div>
</Horizontal>
`;

exports[`<Tag /> rendering when disabled should output correct tree 1`] = `
<Horizontal
  constraint="scale"
>
  <div
    className="wrapper wrapperTypeNormal disabledWrapper"
  >
    <TagNormalBody
      isDisabled={true}
      type="normal"
    >
      Icecream we scream!
    </TagNormalBody>
  </div>
</Horizontal>
`;

exports[`<Tag /> rendering when of type \`normal\` (default) should output correct tree 1`] = `
<Horizontal
  constraint="scale"
>
  <div
    className="wrapper wrapperTypeNormal"
  >
    <TagNormalBody
      isDisabled={false}
      type="normal"
    >
      Icecream we scream!
    </TagNormalBody>
  </div>
</Horizontal>
`;

exports[`<Tag /> rendering when of type \`warning\` should output correct tree 1`] = `
<Horizontal
  constraint="scale"
>
  <div
    className="wrapper wrapperTypeWarning"
  >
    <TagNormalBody
      isDisabled={false}
      type="warning"
    >
      Icecream we scream!
    </TagNormalBody>
  </div>
</Horizontal>
`;

exports[`<Tag /> rendering when removable (\`onRemove\`) should output correct tree 1`] = `
<Horizontal
  constraint="scale"
>
  <div
    className="wrapper wrapperTypeNormal"
  >
    <TagNormalBody
      isDisabled={false}
      onRemove={[MockFunction]}
      type="normal"
    >
      Icecream we scream!
    </TagNormalBody>
    <AccessibleButton
      buttonAttributes={Object {}}
      className="removeWrapper removeWrapperTypeNormal"
      isDisabled={false}
      isToggleButton={false}
      label="Remove"
      onClick={[MockFunction]}
      type="button"
    >
      <CloseBoldIcon
        size="medium"
      />
    </AccessibleButton>
  </div>
</Horizontal>
`;

exports[`\`<TagLinkBody />\` rendering when disabled (\`isDisabled\`) should match the recoreded snapshot 1`] = `
<div
  className="contentWrapper contentWrapperNormal disabledContent"
>
  <TextDetail>
    Icecream we scream!
  </TextDetail>
</div>
`;

exports[`\`<TagLinkBody />\` rendering when enabled (\`isDisabled\`) should match the recoreded snapshot 1`] = `
<div
  className="contentWrapper contentWrapperNormal clickableContentWrapperNormal clickableContentWrapper plainLink"
>
  <Link
    onClick={[MockFunction]}
    replace={false}
    to="foo-too"
  >
    <TextDetail>
      Icecream we scream!
    </TextDetail>
  </Link>
</div>
`;

exports[`\`<TagNormalBody />\` rendering should output correct tree 1`] = `
<div
  className="contentWrapper contentWrapperNormal clickableContentWrapperNormal clickableContentWrapper"
  onClick={[MockFunction]}
>
  <TextDetail>
    Icecream we scream!
  </TextDetail>
</div>
`;
